File: app/src/main/java/com/aandssoftware/aandsinventory/listing/InventoryListAdapter.java
package com.aandssoftware.aandsinventory.listing;

import android.app.SearchManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.SearchView;
import android.widget.SearchView.OnQueryTextListener;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.AppCompatImageButton;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.cardview.widget.CardView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.aandssoftware.aandsinventory.R;
import com.aandssoftware.aandsinventory.common.Utils;
import com.aandssoftware.aandsinventory.database.RealmManager;
import com.aandssoftware.aandsinventory.models.CallbackRealmObject;
import com.aandssoftware.aandsinventory.models.InventoryItem;
import com.aandssoftware.aandsinventory.ui.ListingActivity;
import com.aandssoftware.aandsinventory.ui.OrderDetailsActivity;
import com.aandssoftware.aandsinventory.ui.OrderListActivity;
import com.aandssoftware.aandsinventory.ui.adapters.BaseRealmAdapter.BaseViewHolder;
import io.realm.RealmObject;
import io.realm.RealmResults;

public class InventoryListAdapter implements ListingOperations {
  
  public static final String INVENTORY_ID = "inventory_id";
  private ListingActivity activity;
  
  public class InventoryViewHolder extends BaseViewHolder {
    
    @BindView(R.id.imgInventoryItemHistory)
    ImageView imgInventoryItemHistory;
    @BindView(R.id.imgInventoryItemEdit)
    ImageView imgInventoryItemEdit;
    @BindView(R.id.imgInventoryItemDelete)
    ImageView imgInventoryItemDelete;
    @BindView(R.id.imgInventoryItemLogo)
    ImageView imgInventoryItemLogo;
    @BindView(R.id.inventoryItemName)
    AppCompatTextView inventoryItemName;
    @BindView(R.id.inventoryItemQuantity)
    AppCompatTextView inventoryItemQuantity;
    @BindView(R.id.inventoryItemDetails)
    AppCompatTextView inventoryItemDetails;
    @BindView(R.id.card_view)
    CardView cardView;
    
    @OnClick(R.id.imgInventoryItemHistory)
    public void OnInventoryItemHistory() {
      showInventoryHistory(((InventoryItem) itemView.getTag()).getId());
    }
    
    @OnClick(R.id.imgInventoryItemDelete)
    public void onDeleteClick() {
      deleteInventory((InventoryItem) itemView.getTag(), itemView.getContext());
    }
    
    @OnClick(R.id.imgInventoryItemEdit)
    public void onEditClick() {
      performClick((InventoryItem) itemView.getTag(), itemView, true);
    }
    
    public InventoryViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
  
  public InventoryListAdapter(ListingActivity activity) {
    this.activity = activity;
  }
  
  @Override
  public int getActivityLayoutId() {
    return R.layout.activity_listing;
  }
  
  
  @Override
  public BaseViewHolder getBaseViewHolder(ViewGroup viewGroup, int i) {
    return new InventoryViewHolder(LayoutInflater.from(viewGroup.getContext())
        .inflate(R.layout.inventory_item, viewGroup, false));
  }
  
  @Override
  public void onBindSearchViewHolder(BaseViewHolder baseHolder, int position, RealmObject item) {
    InventoryViewHolder holder = (InventoryViewHolder) baseHolder;
    InventoryItem mItem = (InventoryItem) item;
    holder.inventoryItemName.setText(mItem.getInventoryItemName());
    holder.inventoryItemQuantity.setText(
        mItem.getItemQuantity() + " " + mItem
            .getItemQuantityUnit());
    holder.inventoryItemDetails.setText(
        mItem.getInventoryItemBrandName() + " " + mItem
            .getInventoryItemModelName() + " " + mItem.getInventoryItemColor()
            + " " + mItem.getInventoryItemSize());
    if (mItem.getInventoryItemImagePath() != null) {
      Bitmap bitmap = BitmapFactory.decodeFile(mItem.getInventoryItemImagePath());
      if (null != holder.imgInventoryItemLogo && null != bitmap) {
        holder.imgInventoryItemLogo.setImageBitmap(bitmap);
      }
    }
    holder.cardView.setOnClickListener(new OnClickListener() {
      @Override
      public void onClick(View v) {
        performClick(mItem, holder.itemView, false);
      }
    });
  }
  
  private void performClick(InventoryItem mItem, View itemView, boolean shouldUpdate) {
    if (isOrderSelectionCall()) {
      ShowAddOrderQuantity(mItem, itemView.getContext());
    } else {
      showAddInventoryItemFragment(mItem, shouldUpdate);
    }
  }
  
  private void showInventoryHistory(int id) {
    Intent intent = new Intent(activity, ListingActivity.class);
    intent.putExtra(ListingActivity.LISTING_TYPE, ListType.LIST_TYPE_INVENTORY_HISTORY.ordinal());
    intent.putExtra(InventoryListAdapter.INVENTORY_ID, id);
    activity.startActivityForResult(intent, ListingActivity.LISTING_CODE);
  }
  
  @Override
  public String getTitle() {
    int inventoryType = getInventoryType();
    if (inventoryType == ListType.LIST_TYPE_INVENTORY.ordinal()) {
      return activity.getString(R.string.inventory_item);
    } else {
      return activity.getString(R.string.material);
    }
  }
  
  private boolean isOrderSelectionCall() {
    ComponentName callingActivity = activity.getCallingActivity();
    return (callingActivity != null && callingActivity.getClassName()
        .equalsIgnoreCase(OrderListActivity.class.getName()) || callingActivity.getClassName()
        .equalsIgnoreCase(OrderDetailsActivity.class.getName()));
  }
  
  @Override
  public RealmResults<? extends RealmObject> getResult() {
    int inventoryType = getInventoryType();
    if (inventoryType == ListType.LIST_TYPE_INVENTORY.ordinal()) {
      return RealmManager.getInventoryDao().getInventoryItemRecords();
    } else {
      return RealmManager.getInventoryDao().getMaterialRecords();
    }
  }
  
  public void reloadAdapter(String query) {
    activity.reloadAdapter(
        RealmManager.getInventoryDao().getInventoryItemRecords(getInventoryType(), query));
  }
  
  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    activity.getMenuInflater().inflate(R.menu.inventory_menu, menu);
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
      SearchManager manager = (SearchManager) activity.getSystemService(Context.SEARCH_SERVICE);
      menu.findItem(R.id.search).setVisible(true);
      SearchView search = (SearchView) menu.findItem(R.id.search).getActionView();
      search.setVisibility(View.VISIBLE);
      search.setSearchableInfo(manager.getSearchableInfo(activity.getComponentName()));
      search.setOnQueryTextListener(new OnQueryTextListener() {
        @Override
        public boolean onQueryTextSubmit(String s) {
          return true;
        }
        
        @Override
        public boolean onQueryTextChange(String query) {
          reloadAdapter(query);
          return true;
        }
      });
    }
    return true;
  }
  
  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case android.R.id.home:
        activity.finish();
        return true;
      case R.id.actionAdd:
        showAddInventoryItemFragment(null, false);
        return true;
      default:
    }
    return true;
  }
  
  @Override
  public void onBackPressed() {
    activity.finish();
  }

  public void showAddInventoryItemFragment(InventoryItem mItem, boolean update) {
    AddInventoryFragment fragment = new AddInventoryFragment();
    fragment.setCallbackRealmObject(result -> {
      if (result) {
        Utils.showToast(activity.getString(R.string.inventory_save_message), activity);
        activity.reloadAdapter(getResult());
      }
    });
    fragment.setInventoryItem(mItem);
    fragment.setShouldUpdate(update);
    fragment.show(activity.getFragmentManager(), getTitle());
  }
  
  public void ShowAddOrderQuantity(InventoryItem inventoryItem, Context context) {
    AlertDialog.Builder alertDialogBuilderUserInput =
        new AlertDialog.Builder(context);
    View view = LayoutInflater.from(context).inflate(R.layout.add_order_inventory_item, null);
    alertDialogBuilderUserInput.setView(view);
    setData(view, inventoryItem);
    alertDialogBuilderUserInput
        .setTitle(context.getString(R.string.add_order_inventory_title))
        .setCancelable(false)
        .setPositiveButton(context.getString(R.string.add),
            new DialogInterface.OnClickListener() {
              public void onClick(DialogInterface dialogBox, int id) {
                EditText edtOrderQuantity = ((EditText) view.findViewById(R.id.edtOrderQuantity));
                String newQuantity = edtOrderQuantity.getText().toString();
                int intUpdatedQuantity = Integer.parseInt(inventoryItem.getItemQuantity()) - Integer
                    .parseInt(newQuantity);
                
                InventoryItem mainInventoryCopy = new InventoryItem(inventoryItem);
                mainInventoryCopy.setItemQuantity(String.valueOf(intUpdatedQuantity));
                RealmManager.getInventoryDao().saveInventoryItem(mainInventoryCopy,
                    new CallbackRealmObject() {
                      @Override
                      public void getCallBack(boolean result) {
                      
                      }
                    });
                RealmManager.getCustomerDao()
                    .addInventoryToOrder(mainInventoryCopy, getOrderId(), mainInventoryCopy.getId(),
                        newQuantity,
                        new CallbackRealmObject() {
                          @Override
                          public void getCallBack(boolean result) {
                            Utils.showToast(activity.getString(R.string.inventory_add_order),
                                activity);
                          }
                        });
                dialogBox.cancel();
              }
            })
        .setNegativeButton(
            context.getString(R.string.cancel),
            new DialogInterface.OnClickListener() {
              public void onClick(DialogInterface dialogBox, int id) {
                dialogBox.cancel();
              }
            });
    
    final AlertDialog alertDialog = alertDialogBuilderUserInput.create();
    alertDialog.show();
  }
  
  
  private void setData(View view, InventoryItem inventoryItem) {
    ((EditText) view.findViewById(R.id.edtItemName)).setText(inventoryItem.getInventoryItemName());
    ((EditText) view.findViewById(R.id.edtItemQuantity)).setText(inventoryItem.getItemQuantity());
    ((EditText) view.findViewById(R.id.edtItemUnit)).setText(inventoryItem.getItemQuantityUnit());
    ((EditText) view.findViewById(R.id.edtItemPurchasePrice))
        .setText(inventoryItem.getItemPurchasePrice());
    ((EditText) view.findViewById(R.id.edtItemUnitPrice)).setText(inventoryItem.getItemUnitPrice());
    ((EditText) view.findViewById(R.id.edtSupposedSellingPrice))
        .setText(inventoryItem.getMinimumSellingPrice());
    ((EditText) view.findViewById(R.id.edtItemBrandName))
        .setText(inventoryItem.getInventoryItemBrandName());
    ((EditText) view.findViewById(R.id.edtItemModelName))
        .setText(inventoryItem.getInventoryItemModelName());
    ((EditText) view.findViewById(R.id.edtItemColor))
        .setText(inventoryItem.getInventoryItemColor());
    ((EditText) view.findViewById(R.id.edtItemSize)).setText(inventoryItem.getInventoryItemSize());
    EditText edtOrderQuantity = ((EditText) view.findViewById(R.id.edtOrderQuantity));
    edtOrderQuantity.setText(inventoryItem.getItemQuantity());
    ((AppCompatImageButton) view.findViewById(R.id.addItem)).setOnClickListener(
        new OnClickListener() {
          @Override
          public void onClick(View v) {
            String val = edtOrderQuantity.getText().toString();
            if (!val.isEmpty()) {
              int orderQuantity = Integer.parseInt(val);
              if (orderQuantity > 0) {
                orderQuantity = orderQuantity + 1;
                edtOrderQuantity.setText(String.valueOf(orderQuantity));
              }
            }
          }
        });
    
    ((AppCompatImageButton) view.findViewById(R.id.removeItem)).setOnClickListener(
        new OnClickListener() {
          @Override
          public void onClick(View v) {
            String val = edtOrderQuantity.getText().toString();
            if (!val.isEmpty()) {
              int orderQuantity = Integer.parseInt(val);
              if (orderQuantity > 0) {
                orderQuantity = orderQuantity - 1;
                edtOrderQuantity.setText(String.valueOf(orderQuantity));
              }
            }
          }
        });
  }
  
  public void deleteInventory(InventoryItem inventoryItem, Context context) {
    AlertDialog.Builder alertDialogBuilderUserInput =
        new AlertDialog.Builder(context);
    alertDialogBuilderUserInput
        .setTitle(context.getString(R.string.remove_inventory_item_title))
        .setMessage(context.getString(R.string.remove_inventory_item_message))
        .setCancelable(false)
        .setPositiveButton(context.getString(R.string.yes),
            new DialogInterface.OnClickListener() {
              public void onClick(DialogInterface dialogBox, int id) {
                RealmManager.getInventoryDao().removeInventoryItem(inventoryItem);
                activity.reloadAdapter(getResult());
              }
            })
        .setNegativeButton(
            context.getString(R.string.no),
            new DialogInterface.OnClickListener() {
              public void onClick(DialogInterface dialogBox, int id) {
                dialogBox.cancel();
              }
            });
    
    final AlertDialog alertDialog = alertDialogBuilderUserInput.create();
    alertDialog.show();
  }
  
  private int getInventoryType() {
    return activity.getIntent()
        .getIntExtra(ListingActivity.LISTING_TYPE, ListType.LIST_TYPE_MATERIAL.ordinal());
  }
  
  private int getOrderId() {
    return activity.getIntent()
        .getIntExtra(OrderListActivity.ORDER_ID, -1);
  }

}

File: app/src/main/java/com/aandssoftware/aandsinventory/listing/InventoryListAdapter.kt
package com.aandssoftware.aandsinventory.listing

import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.*
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.SearchView
import android.widget.SearchView.OnQueryTextListener
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.AppCompatTextView
import androidx.cardview.widget.CardView
import com.aandssoftware.aandsinventory.R
import com.aandssoftware.aandsinventory.common.Navigator
import com.aandssoftware.aandsinventory.common.Utils
import com.aandssoftware.aandsinventory.firebase.FirebaseUtil
import com.aandssoftware.aandsinventory.models.InventoryCreatedBy
import com.aandssoftware.aandsinventory.models.InventoryItem
import com.aandssoftware.aandsinventory.models.ViewMode
import com.aandssoftware.aandsinventory.ui.activity.CompanyOrderListActivity
import com.aandssoftware.aandsinventory.ui.activity.ListingActivity
import com.aandssoftware.aandsinventory.ui.activity.OrderDetailsActivity
import com.aandssoftware.aandsinventory.ui.activity.OrderListActivity
import com.aandssoftware.aandsinventory.ui.adapters.BaseAdapter.BaseViewHolder
import com.aandssoftware.aandsinventory.ui.component.CustomEditText
import com.aandssoftware.aandsinventory.utilities.AppConstants
import com.aandssoftware.aandsinventory.utilities.AppConstants.Companion.EMPTY_STRING
import com.aandssoftware.aandsinventory.utilities.AppConstants.Companion.ORDER_RELOAD_LIST_RESULT_CODE
import com.aandssoftware.aandsinventory.utilities.AppConstants.Companion.RELOAD_LIST_RESULT_CODE
import com.aandssoftware.aandsinventory.utilities.CrashlaticsUtil
import com.aandssoftware.aandsinventory.utilities.CrashlaticsUtil.TAG_INFO
import com.bumptech.glide.Glide
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener
import com.google.gson.Gson
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.rxkotlin.flatMapIterable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.inventory_item.view.*
import java.io.Serializable
import java.util.*
import kotlin.collections.ArrayList


class InventoryListAdapter(private val activity: ListingActivity) : ListingOperations {

    private var search: SearchView? = null
    var query: String = EMPTY_STRING
    private val handler: Handler = Handler(Looper.getMainLooper())
    private val workRunnable: Runnable

    companion object {
        const val RECORD_FETCH_AT_TIME = 51
        const val SEARCH_DELAY = 500
    }

    private val isOrderSelectionCall: Boolean
        get() {
            val callingActivity = activity.callingActivity
            return callingActivity != null &&
                    callingActivity.className.equals(OrderListActivity::class.java.name, ignoreCase = true) ||
                    callingActivity!!.className.equals(CompanyOrderListActivity::class.java.name, ignoreCase = true) ||
                    callingActivity.className.equals(OrderDetailsActivity::class.java.name, ignoreCase = true)
        }

    internal var lastNodeKey: Double = 0.0

    private val isAdminUser: Boolean
        get() {
            return Utils.isAdminUser(activity)
        }

    private val inventoryType: Int
        get() = activity.intent
                .getIntExtra(AppConstants.LISTING_TYPE, ListType.LIST_TYPE_MATERIAL.ordinal)

    private val orderId: String?
        get() = activity.intent
                .getStringExtra(AppConstants.ORDER_ID)

    inner class InventoryViewHolder(itemView: View) : BaseViewHolder(itemView) {
        var imgInventoryItemHistory: ImageView = itemView.imgInventoryItemHistory
        var imgInventoryItemEdit: ImageView = itemView.imgInventoryItemEdit
        var imgInventoryItemDelete: ImageView = itemView.imgInventoryItemDelete
        var imgInventoryItemLogo: ImageView = itemView.imgInventoryItemLogo
        var inventoryItemName: AppCompatTextView = itemView.inventoryItemName
        var inventoryItemQuantity: AppCompatTextView = itemView.inventoryItemQuantity
        var inventoryItemDetails: AppCompatTextView = itemView.inventoryItemDetails
        var cardView: CardView = itemView.cardView
        var llButtons: LinearLayout = itemView.llButtons

        init {
            imgInventoryItemHistory.setOnClickListener {
                showInventoryHistory((itemView.tag as InventoryItem).id)
            }
            imgInventoryItemDelete.setOnClickListener {
                var pos: Int = itemView.getTag(R.string.tag) as Int
                deleteInventory(itemView.tag as InventoryItem, itemView.context, pos)
            }
            imgInventoryItemEdit.setOnClickListener {
                var pos: Int = itemView.getTag(R.string.tag) as Int
                performClick(itemView.tag as InventoryItem, ViewMode.UPDATE.ordinal, pos)
            }

            cardView.setOnLongClickListener {
                if (Utils.isAdminUser(activity)) {
                    getDiscountedAmountDialog(itemView.tag as InventoryItem, activity)
                }
                false
            }

            if (Utils.isAdminUser(activity)) {
                llButtons.visibility = View.VISIBLE
            }
        }
    }

    init {
        workRunnable = Runnable { this.performSearch() }
    }

    private fun performSearch() {
        search?.let {
            if (query.isNotEmpty() && it.query.toString().equals(query, ignoreCase = true)) {
                activity.showProgressBar()
                FirebaseUtil.getInstance().getInventoryDao().getAllInventoryItemAtOnce(inventoryType,
                        object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                val list = FirebaseUtil.getInstance()
                                        .getListData(dataSnapshot, InventoryItem::class.java)
                                //val searchedList = ArrayList<InventoryItem>()
                                if (list.isNotEmpty()) {
                                    Observable.just(list)
                                            .flatMapIterable()
                                            .subscribeOn(Schedulers.io())
                                            .observeOn(AndroidSchedulers.mainThread())
                                            .filter {item ->
                                                val searchable = EMPTY_STRING.plus(item.inventoryItemName).plus(item.inventoryItemBrandName).plus(item.inventoryItemModelName).plus(item.description)
                                                var searchableQueryList = query.split(" ", ignoreCase = true)
                                                var add: String? = searchableQueryList.find { content -> searchable.contains(content, ignoreCase = true) }
                                                add?.isNotEmpty()?:false
                                            }
                                            .toList()
                                            .doOnError {
                                                activity.dismissProgressBar()
                                            }
                                            .subscribe {  list->
                                                activity.loadData(ArrayList<InventoryItem>(list))
                                                activity.dismissProgressBar()
                                            }
                                }else{
                                    activity.dismissProgressBar()
                                }
                            }

                            override fun onCancelled(databaseError: DatabaseError) {
                                activity.dismissProgressBar()
                            }
                        })
            }
        }
    }

    override fun getActivityLayoutId(): Int {
        return R.layout.activity_listing
    }


    override fun getBaseViewHolder(viewGroup: ViewGroup, i: Int): BaseViewHolder {
        return InventoryViewHolder(LayoutInflater.from(viewGroup.context)
                .inflate(R.layout.inventory_item, viewGroup, false))
    }

    override fun onBindSearchViewHolder(baseHolder: BaseViewHolder, position: Int, item: Serializable) {
        val holder = baseHolder as InventoryViewHolder
        val mItem = item as InventoryItem
        holder.inventoryItemName.text = mItem.inventoryItemName
        val quantityDetails = EMPTY_STRING.plus(mItem.itemQuantity).plus(" ").plus(mItem.itemQuantityUnit)
        val itemDetails = EMPTY_STRING.plus(mItem.inventoryItemBrandName).plus(" ").plus(mItem
                .inventoryItemModelName).plus(" ").plus(mItem.inventoryItemColor).plus(" ").plus(mItem.inventoryItemSize)

        if (isAdminUser) {
            holder.inventoryItemQuantity.text = quantityDetails
            holder.inventoryItemDetails.text = itemDetails
        } else {
            holder.inventoryItemQuantity.visibility = View.GONE
            holder.inventoryItemDetails.text = itemDetails
        }

        mItem.inventoryItemImagePath?.let {
            if (it.isNotEmpty()) {
                var firstImage = it.values.toMutableList().first()
                var uri: Uri = Uri.parse(firstImage)
                Glide.with(activity)
                        .load(uri)
                        .placeholder(android.R.drawable.ic_menu_gallery)
                        .crossFade()
                        .into(holder.imgInventoryItemLogo)
            }

        }
        holder.cardView.setOnClickListener {
            var pos: Int = baseHolder.itemView.getTag(R.string.tag) as Int
            performClick(mItem, ViewMode.VIEW_ONLY.ordinal, pos)
        }
    }

    private fun performClick(mItem: InventoryItem, viewMode: Int, pos: Int) {
        if (isOrderSelectionCall) {
            addOrderQuantity(mItem, pos)
        } else {
            showAddInventoryItemScreen(mItem.id, viewMode, pos)
        }
    }

    private fun showInventoryHistory(id: String?) {
        val intent = Intent(activity, ListingActivity::class.java)
        intent.putExtra(AppConstants.LISTING_TYPE, ListType.LIST_TYPE_INVENTORY_HISTORY.ordinal)
        intent.putExtra(AppConstants.INVENTORY_ID, id)
        activity.startActivityForResult(intent, AppConstants.LISTING_REQUEST_CODE)
    }

    override fun getTitle(): String = if (inventoryType == ListType.LIST_TYPE_INVENTORY.ordinal) {
        activity.getString(R.string.inventory_item)
    } else {
        activity.getString(R.string.material)
    }

    override fun getResult() {
        activity.showProgressBar()
        val inventoryType = inventoryType
        if (inventoryType == ListType.LIST_TYPE_INVENTORY.ordinal) {
            FirebaseUtil.getInstance().getInventoryDao().getAllInventoryItemAtOnce(lastNodeKey, RECORD_FETCH_AT_TIME,
                    object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val list = FirebaseUtil.getInstance()
                                    .getListData(dataSnapshot, InventoryItem::class.java)
                            if (list.isNotEmpty()) {
                                list.reverse()
                                var shouldLoadMore: Boolean
                                val lastItem = list[list.size - 1]
                                lastItem.let {
                                    lastNodeKey = lastItem.inventoryItemLastUpdatedDate.toDouble()
                                    shouldLoadMore = list.size >= RECORD_FETCH_AT_TIME
                                    if (shouldLoadMore) {
                                        list.remove(lastItem)
                                    }
                                }
                                activity.reloadNewData(list)
                                activity.isLoading = shouldLoadMore.not()
                            }
                            activity.dismissProgressBar()
                        }

                        override fun onCancelled(databaseError: DatabaseError) {
                            activity.dismissProgressBar()
                        }
                    })
        } else {
            FirebaseUtil.getInstance().getInventoryDao().getMaterialRecords(lastNodeKey, RECORD_FETCH_AT_TIME,
                    object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val list = FirebaseUtil.getInstance()
                                    .getListData(dataSnapshot, InventoryItem::class.java)
                            CrashlaticsUtil.logInfo(TAG_INFO, Gson().toJson(list))
                            if (list.isNotEmpty()) {
                                list.reverse()
                                var shouldLoadMore: Boolean
                                val lastItem = list[list.size - 1]
                                lastItem.let {
                                    lastNodeKey = lastItem.inventoryItemLastUpdatedDate.toDouble()
                                    shouldLoadMore = list.size >= RECORD_FETCH_AT_TIME
                                    if (shouldLoadMore) {
                                        list.remove(lastItem)
                                    }
                                }
                                activity.reloadNewData(list)
                                activity.isLoading = shouldLoadMore.not()
                            }
                            activity.dismissProgressBar()
                        }

                        override fun onCancelled(databaseError: DatabaseError) {
                            activity.dismissProgressBar()
                        }
                    })
        }
    }

    fun reloadAdapter(currentQuery: String) {
        query = currentQuery
        handler.removeCallbacks(workRunnable)
        if (currentQuery.isNotEmpty()) {
            handler.postDelayed(workRunnable, SEARCH_DELAY.toLong())
        } else {
            handleEmptyQuery()
        }
    }

    private fun handleEmptyQuery() {
        lastNodeKey = 0.0
        activity.loadData(ArrayList())
        getResult()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        activity.menuInflater.inflate(R.menu.inventory_menu, menu)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
            val manager = activity.getSystemService(Context.SEARCH_SERVICE) as SearchManager
            var menuItem: MenuItem = menu.findItem(R.id.search)
            menuItem.isVisible = true
            search = menu.findItem(R.id.search).actionView as SearchView
            search?.apply {
                this.visibility = View.VISIBLE
                this.setSearchableInfo(manager.getSearchableInfo(activity.componentName))
                this.setOnQueryTextListener(object : OnQueryTextListener {
                    override fun onQueryTextSubmit(s: String): Boolean {
                        return true
                    }

                    override fun onQueryTextChange(query: String): Boolean {
                        reloadAdapter(query)
                        return true
                    }
                })
                this.setOnSearchClickListener(object : View.OnClickListener {
                    override fun onClick(p0: View?) {
                        activity.setActionBarTitleVisibility(View.GONE)
                    }

                })

                this.setOnCloseListener {
                    activity.setActionBarTitleVisibility(View.VISIBLE)
                    false
                }
            }

        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                activity.finish()
                return true
            }
            R.id.actionAdd -> {
                if (Utils.isAdminUser(activity)) {
                    showAddInventoryItemScreen(EMPTY_STRING, ViewMode.ADD.ordinal, AppConstants.INVALID_ID)
                } else {
                    showAddInventoryItemScreen(EMPTY_STRING, ViewMode.ADD_INVENTORY_BY_CUSTOMER.ordinal, AppConstants.INVALID_ID)
                }
                return true
            }
        }
        return true
    }

    override fun onBackPressed() {
        activity.finish()
    }

    private fun showAddInventoryItemScreen(id: String?, viewMode: Int, pos: Int) {
        Navigator.openInventoryScreen(activity, id!!, viewMode, inventoryType, title, pos)
    }

    private fun addOrderQuantity(inventoryItem: InventoryItem, pos: Int) {
        Navigator.openInventoryScreen(activity, inventoryItem.id!!,
                ViewMode.GET_INVENTORY_QUANTITY.ordinal, inventoryType,
                activity.getString(R.string.add_order_inventory_title), orderId!!, pos)
    }


    fun deleteInventory(inventoryItem: InventoryItem, context: Context, pos: Int) {
        val alertDialogBuilderUserInput = AlertDialog.Builder(context)
        alertDialogBuilderUserInput
                .setTitle(context.getString(R.string.remove_inventory_item_title))
                .setMessage(context.getString(R.string.remove_inventory_item_message))
                .setCancelable(false)
                .setPositiveButton(context.getString(R.string.yes)
                ) { dialogBox, id ->
                    FirebaseUtil.getInstance().getInventoryDao()
                            .removeInventoryItem(inventoryItem, inventoryType, DatabaseReference.CompletionListener { databaseError, _ ->
                                if (null == databaseError) {
                                    activity.removeAt(pos)
                                }
                            })
                }
                .setNegativeButton(
                        context.getString(R.string.no)
                ) { dialogBox, id -> dialogBox.cancel() }

        val alertDialog = alertDialogBuilderUserInput.create()
        alertDialog.show()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            AppConstants.GET_INVENTORY_UPDATE_REQUEST_CODE// when we open add inventory screen
            -> when (resultCode) {
                ORDER_RELOAD_LIST_RESULT_CODE//Open add Inventory for getting quantity
                -> callToOrderActivityAfterGettingQuantity(data)
                RELOAD_LIST_RESULT_CODE// added Inventory Item then
                -> callAfterActivityResult(data)
            }
        }
    }

    private fun callAfterActivityResult(data: Intent?) {
        data?.let {
            val viewMode = data.getIntExtra(AppConstants.VIEW_MODE, ViewMode.VIEW_ONLY.ordinal)
            if (viewMode == ViewMode.UPDATE.ordinal) {
                val pos = data.getIntExtra(AppConstants.POSITION_IN_LIST, AppConstants.INVALID_ID)
                if (pos != AppConstants.INVALID_ID) {
                    if (data.hasExtra(AppConstants.INVENTORY_INSTANCE)) {
                        val inventoryInstance: InventoryItem = data.getSerializableExtra(AppConstants.INVENTORY_INSTANCE) as InventoryItem
                        activity.updateElement(inventoryInstance, pos)
                    }
                }
            } else if (viewMode == ViewMode.ADD.ordinal || viewMode == ViewMode.ADD_INVENTORY_BY_CUSTOMER.ordinal) {
                if (data.hasExtra(AppConstants.INVENTORY_INSTANCE)) {
                    val inventoryInstance = data.getSerializableExtra(AppConstants.INVENTORY_INSTANCE)
                    if (inventoryInstance != null) {
                        val inventory: InventoryItem = inventoryInstance as InventoryItem
                        activity.addElement(inventory, 0)
                    }
                }
            }
        }
    }

    private fun callToOrderActivityAfterGettingQuantity(data: Intent?) {
        data?.let {
            var intent = Intent()
            intent.putExtra(AppConstants.UPDATED, true)
            intent.putExtra(AppConstants.POSITION_IN_LIST, activity.intent.getIntExtra(AppConstants.POSITION_IN_LIST, 0))
            intent.putExtras(data)
            activity.setResult(AppConstants.ORDER_RELOAD_LIST_RESULT_CODE, intent)
            activity.finish()
        }
    }

    fun getDiscountedAmountDialog(inventoryItem: InventoryItem, context: Context) {
        val alertDialogBuilderUserInput = AlertDialog.Builder(context)
        var view: View = LayoutInflater.from(activity).inflate(R.layout.customer_discount_dialog, null)
        alertDialogBuilderUserInput
                .setView(view)
                .setCancelable(false)
                .setPositiveButton(context.getString(R.string.yes)) { _, _ ->
                    var text = view.findViewById<CustomEditText>(R.id.edtDiscountedAmount).getText()
                    if (text.isNotEmpty()) {
                        inventoryItem.discountRateForCompany = text.toDouble()
                        var intentReturn = Intent()
                        intentReturn.putExtra(AppConstants.INVENTORY_INSTANCE, inventoryItem)
                        activity.setResult(AppConstants.CUSTOMER_DISCOUNT_RESULT_CODE, intentReturn)
                        activity.finish()
                    } else {
                        activity.showSnackBarMessage(activity.getString(R.string.common_mandatory_error_message))
                    }
                }
                .setNegativeButton(context.getString(R.string.no))
                { dialogBox, _ -> dialogBox.cancel() }

        val alertDialog = alertDialogBuilderUserInput.create()
        alertDialog.show()
    }

}
